File: Lista/Go/Q5.go
package main

import "fmt"

type Animal interface {
    Som() string
}

type Cachorro struct {
    Nome string
}

func (c Cachorro) Som() string {
    return "Au Au"
}

type Gato struct {
    Nome string
}

func (g Gato) Som() string {
    return "Miau"
}

func FazerSons(animais []Animal) {
    for _, animal := range animais {
        fmt.Println(animal.Som())
    }
}

func main() {
    cachorro := Cachorro{Nome: "Rex"}
    gato := Gato{Nome: "Mimi"}

    animais := []Animal{cachorro, gato}

    FazerSons(animais)
}

File: Lista/Python/Q5.py
class Animal:
    def __init__(self, nome):
        self.nome = nome

    def som(self):
        pass

class Cachorro(Animal):
    def som(self):
        return f'{self.nome} faz: Au Au!'

class Gato(Animal):
    def som(self):
        return f'{self.nome} faz: Miau!'

def emitir_sons(animais):
    for animal in animais:
        print(animal.som())

cachorro = Cachorro('Rex')
gato = Gato('Mingau')

animais = [cachorro, gato]

emitir_sons(animais)
